### ISSUE TITLE
ElasticSearchIndexTest failing

### ISSUE THREAD

**Issue by ansell** [eclipse/rdf4j#698]
The ElasticSearchIndexTest is failing for me (as it is for Hudson) in two places with the following messages:

```
Running org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest
Tests run: 5, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 1.453 sec <<< FAILURE! - in org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest
testAddStatement(org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest)  Time elapsed: 0.287 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<6>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest.testAddStatement(ElasticsearchIndexTest.java:141)

testAddMultiple(org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest)  Time elapsed: 0.28 sec  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<7>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest.testAddMultiple(ElasticsearchIndexTest.java:248)


Results :

Failed tests: 
  ElasticsearchIndexTest.testAddStatement:141 expected:<1> but was:<6>
  ElasticsearchIndexTest.testAddMultiple:248 expected:<2> but was:<7>
```

**Comment by ansell** [comment]
A relevant Hudson log is:

https://hudson.eclipse.org/rdf4j/job/rdf4j-master-verify/124/

**Comment by jeenbroekstra** [comment]
Mark, Jacek, either of you available to take a look at this? 


Jeen

**Comment by jeenbroekstra** [comment]
@pulquero, @jgrzebyta

**Comment by jeenbroekstra** [comment]
Let's see if I can reproduce it.


Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: Jeen Broekstra <jeen.broekstra@gmail.com> Date: 19/12/2016  23:01  (GMT+00:00) To: eclipse/rdf4j <reply@reply.github.com> Cc: eclipse/rdf4j <rdf4j@noreply.github.com>, Subscribed <subscribed@noreply.github.com>, Jacek Grzebyta <grzebyta.dev@gmail.com>, Mark Hale <mj_hale@yahoo.com> Subject: Re: [eclipse/rdf4j] ElasticSearchIndexTest failing (#698) 

Mark, Jacek, either of you available to take a look at this? 

Jeen
On 20 Dec 2016, at 09:57, Peter Ansell <notifications@github.com> wrote:
The ElasticSearchIndexTest is failing for me (as it is for Hudson) in two places with the following messages:
Running org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest
Tests run: 5, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 1.453 sec <<< FAILURE! - in org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest
testAddStatement(org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest)  Time elapsed: 0.287 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<6>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest.testAddStatement(ElasticsearchIndexTest.java:141)

testAddMultiple(org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest)  Time elapsed: 0.28 sec  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<7>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchIndexTest.testAddMultiple(ElasticsearchIndexTest.java:248)


Results :

Failed tests: 
  ElasticsearchIndexTest.testAddStatement:141 expected:<1> but was:<6>
  ElasticsearchIndexTest.testAddMultiple:248 expected:<2> but was:<7>
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/eclipse/rdf4j","title":"eclipse/rdf4j","subtitle":"GitHub repository","main_image_url":"<a href="https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png" class="">https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png</a>","avatar_image_url":"<a href="https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png" class="">https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png</a>","action":{"name":"Open in GitHub","url":"<a href="https://github.com/eclipse/rdf4j" class="">https://github.com/eclipse/rdf4j</a>"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"ElasticSearchIndexTest failing (#698)"}],"action":{"name":"View Issue","url":"<a href="https://github.com/eclipse/rdf4j/issues/698" class="">https://github.com/eclipse/rdf4j/issues/698</a>"}}}

**Comment by jeenbroekstra** [comment]
Huh? is this a glitch in GitHub? I'm pretty sure that I didn't send the above mail message....

**Comment by jeenbroekstra** [comment]
Further data point: the tests fail for me locally when I execute `mvn verify` on the project root, from the command line. Executing the test `ElasticsearchSailTest` _directly_ from the IDE, though, they succeed. Very odd.

**Comment by ansell** [comment]
ElasticsearchIndexTest should not be using a single directory for all of its operations "target/test-data", it should be using a JUnit TemporaryFolder Rule. Also the tearDown method in ElasticsearchIndexTest could be improved by using try-finally to reset more thoroughly after that to make sure there are no locks on files inside of the JUnit TemporaryFolder.

**Comment by ansell** [comment]
I added TemporaryFolder and reliably closed/cleaned up but the failure still exists.

I am not familiar with ElasticSearch, but the test seems to be flawed by its reliance on creating a new Node and a new Client separate from those used internally by the ElasticsearchIndex. Any preconditions that the ElasticsearchIndex put on the state for the internal Node and Client may not be transferred accurately to the Node/Client used by the test.

I will open a PR just to fix the TemporaryFolder issue if I cannot fix it quickly.

**Comment by ansell** [comment]
Still can't manage to fix it, but I will open a PR in a second to fix the use of "target/test-data".

I also added a Semaphore to verify that it is an issue within the ElasticSearch APIs and the way we are using them, not in concurrent JUnit test running.

The execution order of the testsuites also do not fit our typical model. The tests in core/sail/fts/elasticsearch/ reference abstract tests in testsuites/. Ie, you have to compile the tests in testsuites before you can compile the module in core/sail/fts/elasticsearch. They need to be moved into compliance modules in the future to match our convention.

**Comment by ansell** [comment]
Well this is strange, I went to commit and there are unexpected files lying around:

```
rdf4j/core/sail/fts/elasticsearch/data/sail_test/nodes/0/indices/elastic-search-sail/
```

It seems like there is a bug somewhere that is writing to a different directory to the configured one.

**Comment by pulquero** [comment]
It is recommended that clients have their own (non-data storing) nodes to handle routing to data nodes in the cluster rather than add that burden to data nodes. The cluster should be in a consistent state in the test because it explicitly waits for the overall cluster health to be green. 

    On Tuesday, 20 December 2016, 1:58, Peter Ansell <notifications@github.com> wrote:
 

 I added TemporaryFolder and reliably closed/cleaned up but the failure still exists.I am not familiar with ElasticSearch, but the test seems to be flawed by its reliance on creating a new Node and a new Client separate from those used internally by the ElasticsearchIndex. Any preconditions that the ElasticsearchIndex put on the state for the internal Node and Client may not be transferred accurately to the Node/Client used by the test.I will open a PR just to fix the TemporaryFolder issue if I cannot fix it quickly.—

**Comment by pulquero** [comment]
I'm wondering if the problem is related to the addition of the new AbstractLuceneSailSpinTest and AbstractLuceneSailTupleFunctionTest because previously we haven't seen this issue. Note that from the log, ElasticsearchIndexTest runs after both of them. I'm gonna see if disabling those two tests has any influence on the failures.

 

    On Tuesday, 20 December 2016, 2:29, Peter Ansell <notifications@github.com> wrote:
 

 Well this is strange, I went to commit and there are unexpected files lying around:rdf4j/core/sail/fts/elasticsearch/data/sail_test/nodes/0/indices/elastic-search-sail/
It seems like there is a bug somewhere that is writing to a different directory to the configured one.—

**Comment by pulquero** [comment]
When I run maven against master I get:

> Tests run: 18, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.695 sec <<< FAILURE! - in org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchSailTest
testContextHandling(org.eclipse.rdf4j.sail.elasticsearch.ElasticsearchSailTest)  Time elapsed: 0.258 sec  <<< FAILURE!
java.lang.AssertionError: query for literal 'sfourponecone did not return the expected resource expected:<urn:subject4> but was:<urn:subject1>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.eclipse.rdf4j.sail.lucene.AbstractLuceneSailTest.assertQueryResult(AbstractLuceneSailTest.java:1099)
        at org.eclipse.rdf4j.sail.lucene.AbstractLuceneSailTest.testContextHandling(AbstractLuceneSailTest.java:853)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:745)

**Comment by jeenbroekstra** [comment]
@ansell what's the status of this? I haven't seen this come up lately, has this been resolved and/or can we close the issue?

**Comment by ansell** [comment]
I think this was fixed in #700 as I haven't seen it since then.
